src/docker.rs

```diff
@@ -343,9 +755,302 @@ pub fn run(
         docker.args(&["-w".as_ref(), mount_wd.as_os_str()]);
     }
 
+    Ok(())
+}
+
+#[allow(clippy::too_many_arguments)] // TODO: refactor
+fn remote_run(
+    target: &Target,
+    args: &[String],
+    metadata: &CargoMetadata,
+    config: &Config,
+    uses_xargo: bool,
+    sysroot: &Path,
+    verbose: bool,
+    docker_in_docker: bool,
+    cwd: &Path,
+) -> Result<ExitStatus> {
+    let dirs = Directories::create(metadata, cwd, sysroot, docker_in_docker, verbose)?;
+
+    let mut cmd = cargo_cmd(uses_xargo);
+    cmd.args(args);
+
+    let engine = Engine::new(true, verbose)?;
+    let mount_prefix = "/cross";
+
+    // the logic is broken into the following steps
+    // 1. get our unique identifiers and cleanup from a previous run.
+    // 2. create a data volume to store everything
+    // 3. start our container with the data volume and all envvars
+    // 4. copy all mounted volumes over
+    // 5. create symlinks for all mounted data
+    // 6. execute our cargo command inside the container
+    // 7. copy data from target dir back to host
+    // 8. stop container and delete data volume
+    //
+    // we use structs that wrap the resources to ensure they're dropped
+    // in the correct order even on error, to ensure safe cleanup
+
+    // 1. get our unique identifiers and cleanup from a previous run.
+    // this can happen if we didn't gracefully exit before
+    let container = remote_identifier(target, metadata, &dirs)?;
+    let volume = VolumeId::create(&engine, &container, verbose)?;
+    let state = container_state(&engine, &container, verbose)?;
+    if !state.is_stopped() {
+        eprintln!("warning: container {container} was running.");
+        container_stop(&engine, &container, verbose)?;
+    }
+    if state.exists() {
+        eprintln!("warning: container {container} was exited.");
+        container_rm(&engine, &container, verbose)?;
+    }
+    if let VolumeId::Discard(ref id) = volume {
+        if volume_exists(&engine, id, verbose)? {
+            eprintln!("warning: temporary volume {container} existed.");
+            volume_rm(&engine, id, verbose)?;
+        }
+    }
+
+    // 2. create our volume to copy all our data over to
+    if let VolumeId::Discard(ref id) = volume {
+        volume_create(&engine, id, verbose)?;
+    }
+    let _volume_deletter = DeleteVolume(&engine, &volume, verbose);
+
+    // 3. create our start container command here
+    let mut docker = docker_subcommand(&engine, "run");
+    docker.args(&["--userns", "host"]);
```

Comment:
Probably off topic but how is `userns=host` beneficial? https://docs.docker.com/engine/security/userns-remap/#disable-namespace-remapping-for-a-container didn't clear things up for me.

Reply:
There's a few reasons, and this has been around since 2017. Some of them are historical with our design (I'm a relatively new maintainer, so I might not know everything), but the reasons are partially documented in #90. Our android images used to require the `--privileged` flag, which is incompatible with userns remapping (until #746, where we removed this need and mitigated the permissions a container could use). It also interferes with bind mounts, which we make extensive use of. 

Finally, we remap the user in the container to the current user anyway using the `--user 1000:1000` flag (normally, if run as root it can be `0:0` I believe), so this really isn't an issue anyway. Stuff run as `--user 1000:1000` means it's a generic, unprivileged user so the permissions and owner of the files generated will be the same as your current user, but also the commands will be run unprivileged. This means user namespace remapping isn't that useful for us anyway.

I hope that explains it? Also, in general, anyone using cross has access to a Docker or Podman daemon, which in the former probably has root permissions (unless using rootless docker) on the machine, so it's really only a concern if you're worried the crate's build system is dangerous (a valid concern), but we have some mitigations in place for that anyway.